@xen-orchestra/xapi/src/vm.js

```diff
@@ -300,25 +299,22 @@ module.exports = class Vm {
       ignoreErrors.call(asyncMap(vm.snapshots, _ => this.VM_destroy(_))),
       deleteDisks &&
         ignoreErrors.call(
-          asyncMap(disks, vdiRef =>
-            pRetry(
-              async () => {
-                // list VMs connected to this VDI
-                const vmRefs = await asyncMap(await this.getField('VDI', vdiRef, 'VBDs'), vbdRef =>
-                  this.getField('VBD', vbdRef, 'VM')
-                )
-                if (vmRefs.every(_ => _ === vmRef)) {
-                  return this.callAsync('VDI.destroy', vdiRef)
-                }
-                throw new AttachedVdiError()
-              },
-              {
-                delay: 5e3,
-                tries: 2,
-                when: AttachedVdiError,
+          asyncMap(disks, async vdiRef => {
+            // Dont destroy if attached to other (non control domain) VMs
+            for (const vbdRef of await this.getField('VDI', vdiRef, 'VBDs')) {
+              const vmRef2 = await this.getField('VBD', vbdRef, 'VM')
+              if (vmRef2 !== vmRef && !(await this.getField('VM', vmRef, 'is_control_domain'))) {
+                return
               }
-            )
-          )
+            }
+
+            return pRetry(() => this.VDI_destroy(vdiRef), {
+              // work around a race condition in XCP-ng/XenServer where the disk is not fully unmounted yet.
+              delay: 5e3,
+              retries: 5,
```

Comment:
I think I would sleep better with 10 retries.

Reply:
I'll make it configurable :slightly_smiling_face: